src/dmd/backend/dlist.d

```diff
@@ -0,0 +1,570 @@
+        {   if (list2)
+                result = -1;    /* list1 < list2        */
+            break;
+        }
+        if (!list2)
+        {   result = 1;         /* list1 > list2        */
+            break;
+        }
+        result = (*fp)(list_ptr(list1),list_ptr(list2));
+        if (result)
+            break;
+        list1 = list_next(list1);
+        list2 = list_next(list2);
+    }
+    return result;
+}
+
+/*************************
+ * Search for ptr in list.
+ * Returns:
+ *    If found, return list entry that it is, else null.
+ */
+
+list_t list_inlist(list_t list, void* ptr)
+{
+    for (; list; list = list_next(list))
+        if (list.ptr == ptr)
+            break;
+    return list;
+}
+
+/*************************
+ * Concatenate two lists (l2 appended to l1).
+ * Output:
+ *      *pl1 updated to be start of concatenated list.
+ * Returns:
+ *      *pl1
+ */
+
+list_t list_cat(list_t *pl1, list_t l2)
+{
+    list_t *pl;
+    for (pl = pl1; *pl; pl = &(*pl).next)
+    { }
+    *pl = l2;
+    return *pl1;
+}
+
+/***************************************
+ * Apply a function fp to each member of a list.
+ */
+
+void list_apply(list_t* plist, void function(void*) @nogc nothrow fp)
+{
+    if (fp)
+        for (list_t l = *plist; l; l = list_next(l))
+        {
+            (*fp)(list_ptr(l));
+        }
+}
+
+/********************************************
+ * Reverse a list in place.
+ */
+
+list_t list_reverse(list_t l)
+{
+    list_t r = null;
+    while (l)
+    {
+        list_t ln = list_next(l);
+        l.next = r;
+        r = l;
+        l = ln;
+    }
+    return r;
+}
+
+
+/**********************************************
+ * Copy list of pointers into an array of pointers.
+ */
+
+void list_copyinto(list_t l, void *pa)
+{
+    void **ppa = cast(void **)pa;
+    for (; l; l = l.next)
+    {
+        *ppa = l.ptr;
+        ++ppa;
+    }
+}
+
+/**********************************************
+ * Insert item into list at nth position.
+ */
+
+list_t list_insert(list_t *pl,void *ptr,int n)
+{
+    list_t list;
+
+    while (n)
+    {
+        pl = &(*pl).next;
+        n--;
+    }
+    list = list_alloc();
+    if (list)
+    {
+        list.next = *pl;
+        *pl = list;
+        list.ptr = ptr;
+        list.count = 1;
+    }
+    return list;
+}
+
+}
+
+/* The following function should be nothrow @nogc, too, but on
+ * some platforms core.stdc.stdarg is not fully nothrow @nogc.
+ */
+
+/*************************
+ * Build a list out of the null-terminated argument list.
+ * Returns:
+ *      generated list
+ */
+
+list_t list_build(void *p,...)
+{
+    va_list ap;
+
+    list_t alist = null;
+    list_t *pe = &alist;
+    for (va_start(ap,p); p; p = va_arg!(void*)(ap))
+    {
+        list_t list = list_alloc();
+        if (list)
+        {
+            list.next = null;
+            list.ptr = p;
+            list.count = 1;
+            *pe = list;
+            pe = &list.next;
+        }
+    }
+    //va_end(ap);
```

Comment:
What's the reason for not calling `va_end` here? 
(in C it's called)

Reply:
There was some problem with it being @nogc and nothrow. core.stdc.stdarg apparently needs some additional work.

Reply:
Compiles fine for me -> https://github.com/dlang/dmd/pull/8367